File: exercicios/aula13/ex07_comandos/comandos.py
"""Exercício 07 — ações semânticas para comandos e fluxo de controle.

Constrói a AST de comandos de um subconjunto MINIC (sem gerar código). As
expressões vêm do parser do exercício 06 (ex06_expressoes_minic/expr.py).

Gramática:

    programa -> comando*
    comando  -> bloco | if | while | return | atrib | expr_cmd
    bloco    -> { comando* }
    if       -> if ( expr ) comando [ else comando ]
    while    -> while ( expr ) comando
    return   -> return [ expr ] ;
    atrib    -> id = expr ;
    expr_cmd -> expr ;

Ações (atributo sintetizado `node`) e invariantes de cada nó:

    bloco    Block(comandos)          lista na ordem do texto; pode ser vazia
    if       If(cond, entao, senao)   cond e entao nunca ausentes; senao = None
                                      quando não há else (ausência explícita)
    while    While(cond, corpo)       os dois sempre presentes
    return   Return(valor)            valor = None em `return;`
    atrib    Assign(alvo, valor)      alvo é sempre um Identifier
    expr_cmd ExprStmt(expr)

`atrib` e `expr_cmd` começam igual (uma expressão). O parser lê a expressão e,
se vier `=`, exige que ela seja um identificador — só então o nó vira Assign.
O `else` pendente fica com o `if` mais próximo, porque o `if` interno consome
o `else` assim que o encontra.

Uso:  python comandos.py <arquivo | ->
"""

import sys
from dataclasses import dataclass, field
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "ex06_expressoes_minic"))
import expr as E  # noqa: E402
from expr import F  # noqa: E402


@dataclass(frozen=True)
class Block:
    comandos: tuple
    linha: int
    coluna: int


@dataclass(frozen=True)
class Assign:
    alvo: E.Identifier
    valor: E.Expr
    linha: int
    coluna: int


@dataclass(frozen=True)
class If:
    cond: E.Expr
    entao: "Comando"
    senao: "Comando | None"
    linha: int
    coluna: int


@dataclass(frozen=True)
class While:
    cond: E.Expr
    corpo: "Comando"
    linha: int
    coluna: int


@dataclass(frozen=True)
class Return:
    valor: E.Expr | None
    linha: int
    coluna: int


@dataclass(frozen=True)
class ExprStmt:
    expr: E.Expr
    linha: int
    coluna: int


Comando = Block | Assign | If | While | Return | ExprStmt


class Parser:
    def __init__(self, fl: F.Fluxo):
        self.fl = fl
        self.expr = E.ExprParser(fl)

    def parse_programa(self) -> list:
        comandos = []
        while not self.fl.verifica("EOF"):
            comandos.append(self.parse_comando())
        return comandos

    def parse_comando(self) -> Comando:
        tok = self.fl.atual()
        if tok.type == "LBRACE":
            return self.parse_bloco()
        if tok.type == "KW_IF":
            self.fl.proximo()
            cond = self._condicao()
            entao = self.parse_comando()
            senao = self.parse_comando() if self.fl.aceita("KW_ELSE") else None
            return If(cond, entao, senao, tok.line, tok.column)
        if tok.type == "KW_WHILE":
            self.fl.proximo()
            cond = self._condicao()
            return While(cond, self.parse_comando(), tok.line, tok.column)
        if tok.type == "KW_RETURN":
            self.fl.proximo()
            valor = None if self.fl.verifica("SEMICOLON") else self.expr.parse_expr()
            self.fl.exige("SEMICOLON", "operador ou ';'")
            return Return(valor, tok.line, tok.column)
        if tok.type in ("RBRACE", "KW_ELSE", "EOF", "RPAREN", "SEMICOLON"):
            raise F.ErroSintaxe(tok, "comando")
        e = self.expr.parse_expr()
        igual = self.fl.aceita("ASSIGN")
        if igual is not None:
            if not isinstance(e, E.Identifier):
                raise F.ErroSintaxe(igual, "';' (o lado esquerdo de '=' deve ser um identificador)")
            valor = self.expr.parse_expr()
            self.fl.exige("SEMICOLON", "operador ou ';'")
            return Assign(e, valor, e.linha, e.coluna)
        self.fl.exige("SEMICOLON", "operador, '=' ou ';'")
        return ExprStmt(e, tok.line, tok.column)

    def parse_bloco(self) -> Block:
        abre = self.fl.exige("LBRACE", "'{'")
        comandos = []
        while not self.fl.verifica("RBRACE"):
            if self.fl.verifica("EOF"):
                raise F.ErroSintaxe(self.fl.atual(), "'}'")
            comandos.append(self.parse_comando())
        self.fl.proximo()
        return Block(tuple(comandos), abre.line, abre.column)

    def _condicao(self) -> E.Expr:
        self.fl.exige("LPAREN", "'('")
        cond = self.expr.parse_expr()
        self.fl.exige("RPAREN", "operador ou ')'")
        return cond


def imprime(no: Comando, nivel: int, rotulo: str = ""):
    ind = "  " * nivel
    p = f"{ind}{rotulo}: " if rotulo else ind
    filho = "  " * (nivel + 1)
    match no:
        case Block(comandos=cs, linha=l, coluna=c):
            print(f"{p}Block @{l}:{c}" + (" (vazio)" if not cs else ""))
            for cmd in cs:
                imprime(cmd, nivel + 1)
        case Assign(alvo=a, valor=v, linha=l, coluna=c):
            print(f"{p}Assign @{l}:{c}")
            print(f"{filho}alvo: {E.sexpr(a)}")
            print(f"{filho}valor: {E.sexpr(v)}")
        case If(cond=cond, entao=entao, senao=senao, linha=l, coluna=c):
            print(f"{p}If @{l}:{c}")
            print(f"{filho}cond: {E.sexpr(cond)}")
            imprime(entao, nivel + 1, "entao")
            if senao is None:
                print(f"{filho}senao: <ausente>")
            else:
                imprime(senao, nivel + 1, "senao")
        case While(cond=cond, corpo=corpo, linha=l, coluna=c):
            print(f"{p}While @{l}:{c}")
            print(f"{filho}cond: {E.sexpr(cond)}")
            imprime(corpo, nivel + 1, "corpo")
        case Return(valor=v, linha=l, coluna=c):
            print(f"{p}Return @{l}:{c}")
            print(f"{filho}valor: " + ("<ausente>" if v is None else E.sexpr(v)))
        case ExprStmt(expr=e, linha=l, coluna=c):
            print(f"{p}ExprStmt @{l}:{c}")
            print(f"{filho}expr: {E.sexpr(e)}")


def main(argv: list[str]) -> int:
    F.configurar_saida()
    if len(argv) != 2:
        print("uso: python comandos.py <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO
    fl, codigo = F.abrir(argv[1])
    if fl is None:
        return codigo
    try:
        comandos = Parser(fl).parse_programa()
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE
    print("Programa" + (" (vazio)" if not comandos else ""))
    for cmd in comandos:
        imprime(cmd, 1)
    return F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: exercicios/aula13/ex07_comandos/comandos.c
/* Exercício 07 — ações semânticas para comandos e fluxo de controle.
 *
 * Mesma gramática, mesmos nós, mesmas invariantes e mesma saída de
 * comandos.py (ver a docstring de lá). As expressões vêm de
 * ex06_expressoes_minic/expr.c.
 *
 * Em C, `Cmd` é uma struct discriminada por `CmdKind`, com um construtor por
 * comando. Posse: cada Cmd é dono das suas expressões e dos seus comandos
 * filhos; cmd_free destrói a subárvore recursivamente. Em erro de sintaxe, cada
 * função de parse libera o que já construiu antes de devolver NULL.
 *
 * Uso:  ./comandos <arquivo | ->
 */

#include <stdio.h>
#include <stdlib.h>

#include "../ex06_expressoes_minic/expr.h"

typedef enum { C_BLOCK, C_ASSIGN, C_IF, C_WHILE, C_RETURN, C_EXPR } CmdKind;

typedef struct Cmd {
    CmdKind k;
    int linha, coluna;
    Expr *e1;          /* ASSIGN: alvo; IF/WHILE: cond; RETURN: valor (ou NULL); EXPR: expr */
    Expr *e2;          /* ASSIGN: valor */
    struct Cmd *c1;    /* IF: entao; WHILE: corpo */
    struct Cmd *c2;    /* IF: senao (NULL = sem else) */
    struct Cmd **lista; /* BLOCK: comandos, na ordem do texto */
    size_t n, cap;
} Cmd;

static void cmd_free(Cmd *c) {
    if (!c) return;
    expr_free(c->e1);
    expr_free(c->e2);
    cmd_free(c->c1);
    cmd_free(c->c2);
    for (size_t i = 0; i < c->n; i++) cmd_free(c->lista[i]);
    free(c->lista);
    free(c);
}

/* -- construtores --------------------------------------------------------- */

static Cmd *cmd_novo(CmdKind k, int linha, int coluna) {
    Cmd *c = aloca(sizeof(Cmd));
    *c = (Cmd){.k = k, .linha = linha, .coluna = coluna};
    return c;
}

static void cmd_block_add(Cmd *bloco, Cmd *c) {
    if (bloco->n == bloco->cap) {
        bloco->cap = bloco->cap ? bloco->cap * 2 : 4;
        bloco->lista = realoca(bloco->lista, bloco->cap * sizeof(Cmd *));
    }
    bloco->lista[bloco->n++] = c;
}

static Cmd *cmd_assign(Expr *alvo, Expr *valor) {
    Cmd *c = cmd_novo(C_ASSIGN, alvo->linha, alvo->coluna);
    c->e1 = alvo, c->e2 = valor;
    return c;
}

static Cmd *cmd_if(const Token *t, Expr *cond, Cmd *entao, Cmd *senao) {
    Cmd *c = cmd_novo(C_IF, t->line, t->column);
    c->e1 = cond, c->c1 = entao, c->c2 = senao;
    return c;
}

static Cmd *cmd_while(const Token *t, Expr *cond, Cmd *corpo) {
    Cmd *c = cmd_novo(C_WHILE, t->line, t->column);
    c->e1 = cond, c->c1 = corpo;
    return c;
}

static Cmd *cmd_return(const Token *t, Expr *valor) {
    Cmd *c = cmd_novo(C_RETURN, t->line, t->column);
    c->e1 = valor;
    return c;
}

static Cmd *cmd_expr(const Token *t, Expr *e) {
    Cmd *c = cmd_novo(C_EXPR, t->line, t->column);
    c->e1 = e;
    return c;
}

/* -- parser ---------------------------------------------------------------- */

static int exige(Fluxo *fl, TokenType tipo, const char *esperado) {
    if (fluxo_aceita(fl, tipo)) return 1;
    erro_sintaxe(fluxo_atual(fl), esperado);
    return 0;
}

static Expr *expressao(Fluxo *fl) {
    ExprErro erro;
    Expr *e = expr_parse(fl, &erro);
    if (!e) erro_sintaxe(erro.token, erro.esperado);
    return e;
}

static Expr *condicao(Fluxo *fl) {
    if (!exige(fl, TOK_LPAREN, "'('")) return NULL;
    Expr *e = expressao(fl);
    if (e && !exige(fl, TOK_RPAREN, "operador ou ')'")) {
        expr_free(e);
        return NULL;
    }
    return e;
}

static Cmd *parse_comando(Fluxo *fl);

static Cmd *parse_bloco(Fluxo *fl) {
    const Token *abre = fluxo_proximo(fl); /* '{' */
    Cmd *b = cmd_novo(C_BLOCK, abre->line, abre->column);
    while (!fluxo_verifica(fl, TOK_RBRACE)) {
        if (fluxo_verifica(fl, TOK_EOF)) {
            erro_sintaxe(fluxo_atual(fl), "'}'");
            cmd_free(b);
            return NULL;
        }
        Cmd *c = parse_comando(fl);
        if (!c) {
            cmd_free(b);
            return NULL;
        }
        cmd_block_add(b, c);
    }
    fluxo_proximo(fl);
    return b;
}

static Cmd *parse_comando(Fluxo *fl) {
    const Token *t = fluxo_atual(fl);
    switch (t->type) {
    case TOK_LBRACE: return parse_bloco(fl);
    case TOK_KW_IF: {
        fluxo_proximo(fl);
        Expr *cond = condicao(fl);
        if (!cond) return NULL;
        Cmd *entao = parse_comando(fl);
        if (!entao) {
            expr_free(cond);
            return NULL;
        }
        Cmd *senao = NULL;
        if (fluxo_aceita(fl, TOK_KW_ELSE) && !(senao = parse_comando(fl))) {
            expr_free(cond), cmd_free(entao);
            return NULL;
        }
        return cmd_if(t, cond, entao, senao);
    }
    case TOK_KW_WHILE: {
        fluxo_proximo(fl);
        Expr *cond = condicao(fl);
        if (!cond) return NULL;
        Cmd *corpo = parse_comando(fl);
        if (!corpo) {
            expr_free(cond);
            return NULL;
        }
        return cmd_while(t, cond, corpo);
    }
    case TOK_KW_RETURN: {
        fluxo_proximo(fl);
        Expr *valor = NULL;
        if (!fluxo_verifica(fl, TOK_SEMICOLON) && !(valor = expressao(fl))) return NULL;
        if (!exige(fl, TOK_SEMICOLON, "operador ou ';'")) {
            expr_free(valor);
            return NULL;
        }
        return cmd_return(t, valor);
    }
    case TOK_RBRACE:
    case TOK_KW_ELSE:
    case TOK_EOF:
    case TOK_RPAREN:
    case TOK_SEMICOLON: erro_sintaxe(t, "comando"); return NULL;
    default: break;
    }
    Expr *e = expressao(fl);
    if (!e) return NULL;
    if (fluxo_verifica(fl, TOK_ASSIGN)) {
        const Token *igual = fluxo_proximo(fl);
        if (e->k != E_IDENT) {
            erro_sintaxe(igual, "';' (o lado esquerdo de '=' deve ser um identificador)");
            expr_free(e);
            return NULL;
        }
        Expr *valor = expressao(fl);
        if (!valor) {
            expr_free(e);
            return NULL;
        }
        if (!exige(fl, TOK_SEMICOLON, "operador ou ';'")) {
            expr_free(e), expr_free(valor);
            return NULL;
        }
        return cmd_assign(e, valor);
    }
    if (!exige(fl, TOK_SEMICOLON, "operador, '=' ou ';'")) {
        expr_free(e);
        return NULL;
    }
    return cmd_expr(t, e);
}

/* -- impressão ------------------------------------------------------------- */

static void linha_expr(int nivel, const char *rotulo, const Expr *e) {
    printf("%*s%s: ", 2 * nivel, "", rotulo);
    if (e) expr_sexpr(stdout, e, 0);
    else fputs("<ausente>", stdout);
    putchar('\n');
}

static void imprime(const Cmd *c, int nivel, const char *rotulo) {
    printf("%*s", 2 * nivel, "");
    if (rotulo) printf("%s: ", rotulo);
    switch (c->k) {
    case C_BLOCK:
        printf("Block @%d:%d%s\n", c->linha, c->coluna, c->n ? "" : " (vazio)");
        for (size_t i = 0; i < c->n; i++) imprime(c->lista[i], nivel + 1, NULL);
        break;
    case C_ASSIGN:
        printf("Assign @%d:%d\n", c->linha, c->coluna);
        linha_expr(nivel + 1, "alvo", c->e1);
        linha_expr(nivel + 1, "valor", c->e2);
        break;
    case C_IF:
        printf("If @%d:%d\n", c->linha, c->coluna);
        linha_expr(nivel + 1, "cond", c->e1);
        imprime(c->c1, nivel + 1, "entao");
        if (c->c2) imprime(c->c2, nivel + 1, "senao");
        else printf("%*ssenao: <ausente>\n", 2 * (nivel + 1), "");
        break;
    case C_WHILE:
        printf("While @%d:%d\n", c->linha, c->coluna);
        linha_expr(nivel + 1, "cond", c->e1);
        imprime(c->c1, nivel + 1, "corpo");
        break;
    case C_RETURN:
        printf("Return @%d:%d\n", c->linha, c->coluna);
        linha_expr(nivel + 1, "valor", c->e1);
        break;
    case C_EXPR:
        printf("ExprStmt @%d:%d\n", c->linha, c->coluna);
        linha_expr(nivel + 1, "expr", c->e1);
        break;
    }
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "uso: %s <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, argv[1]);
    if (codigo != OK) return codigo;

    Cmd *programa = cmd_novo(C_BLOCK, 0, 0); /* só um contêiner para a lista */
    while (codigo == OK && !fluxo_verifica(&fl, TOK_EOF)) {
        Cmd *c = parse_comando(&fl);
        if (c) cmd_block_add(programa, c);
        else codigo = ERRO_SINTAXE;
    }
    if (codigo == OK) {
        printf("Programa%s\n", programa->n ? "" : " (vazio)");
        for (size_t i = 0; i < programa->n; i++) imprime(programa->lista[i], 1, NULL);
    }
    cmd_free(programa);
    fluxo_fechar(&fl);
    return codigo;
}
